OSSM-8268: Getting traffic into the mesh (#83901)
openshift/openshift-docs: modules/ossm-about-configuring-a-gateway-to-accept-ingress-traffic.adoc

// Module included in the following assemblies:

// gateways/ossm-about-gateways.adoc

:_mod-docs-content-type: Concept
[id="ossm-about-configuring-a-gateway-to-accept-ingress-traffic_{context}"]
= About configuring a gateway to accept ingress traffic
:context: ossm-about-configuring-a-gateway-to-accept-ingress-traffic

> When you install a gateway using gateway injection you can configure it to receive ingress traffic using both the {istio} `Gateway` and `VirtualService` resources. The {istio} `Gateway` resource describes a load balancer operating at the edge of the mesh receiving incoming or outgoing HTTP/TCP connections. The `Gateway` specification describes a set of ports that should be exposed, the type of protocol to use, and the Server Name Indication (SNI) configuration for the load balancer. `VirtualServices` are used to define routing rules that apply to an {istio} `Gateway`, similar to how you can define routing rules for internal mesh traffic.

Author: Done. LEt's look at this though as I made a small tweak.

> When you install a gateway using gateway injection you can configure it to receive ingress traffic using both the {istio} `Gateway` and `VirtualService` resources. The {istio} `Gateway` resource describes a load balancer operating at the edge of the mesh receiving incoming or outgoing HTTP/TCP connections. The `Gateway` specification describes a set of ports that should be exposed, the type of protocol to use, and the Server Name Indication (SNI) configuration for the load balancer. `VirtualServices` are used to define routing rules that apply to an {istio} `Gateway`, similar to how you can define routing rules for internal mesh traffic.

Author: we chatted. done.

> 

Reviewer: Deploy a bookinfo application into a mesh namespace.

oc create ns bookinfo 
oc label namespace bookinfo istio-injection=enabled
 oc apply -f https://raw.githubusercontent.com/istio/istio/release-1.23/samples/bookinfo/platform/kube/bookinfo.yaml -n bookinfo

Author: We discussed and don't think we need this information in this topic.

> In the following example an Istio Gateway resource configures a gateway proxy to act as a load balancer. This configuration exposes port 443 (HTTPS) for the host, `bookinfo.com`. The configuration will apply to gateway pods in any namespace with the labeled `istio: ingressgateway`.

Author: done

> In the following example an {istio} Gateway resource configures a gateway proxy to act as a load balancer. This configuration exposes port 443 (HTTPS) for the host, `bookinfo.com`. The configuration applies in all namespaces for gateway pods with the labeled `istio: ingressgateway`.

Reviewer 2: ```suggestion
In the following example an {istio} Gateway resource configures a gateway proxy to act as an entry point for external traffic. This configuration exposes port 443 (HTTPS) for the host, `bookinfo.com`. The configuration applies to pods labelled with `istio: ingressgateway` which typically correspond to ingress gateway pods. The tls mode is configured as SIMPLE, so it would terminate the incoming HTTPS traffic using the provided certificate and private key.
```

Author: done

.Sample configuration
[source,yaml,subs="attributes,verbatim"]
----
apiVersion: networking.istio.io/v1
kind: Gateway
metadata:
  name: bookinfo-gateway
  namespace: bookinfo
spec:
  selector:

>     istio: ingressgateway

Reviewer: istio: ingressgateway replace with the   istio: <gateway_name>

Author: done

>     istio: <gateway_name>

Reviewer 2: To be aligned with the statement in Line 12 above.

```suggestion
    istio: ingressgateway
```

Author: done

  servers:
  - port:
      number: 443
      name: https-443
      protocol: HTTPS
    hosts:
    - bookinfo.com
    tls:
      mode: SIMPLE
      serverCertificate: /etc/certs/servercert.pem
      privateKey: /etc/certs/privatekey.pem
----

> The following `VirtualService` is bound to the {istio} `Gateway` resource previously shown. The specification defines rules to route traffic with the `/reviews/` path prefix to the reviews service in the `bookinfo` namespace.

Reviewer 2: ```suggestion
The following `VirtualService` is bound to the {istio} `Gateway` resource previously shown. The specification defines rules to route traffic with the `/reviews/` path prefix to the reviews service in the `bookinfo` namespace. The `VirtualService` explicitly references the gateway to ensure that the rules are only applied to traffic entering through the specified gateway.
```

Author: done

.Sample configuration
[source,yaml,subs="attributes,verbatim"]
----
kind: VirtualService
metadata:
  name: bookinfo-rule
  namespace: bookinfo
spec:
  hosts:
  - bookinfo.com
  gateways:

>   - <gateway_resource_namespace>/bookinfo-gateway

Reviewer 2: ```suggestion
  - bookinfo/bookinfo-gateway
```

Author: done

> The following `VirtualService` is bound to the {istio} `Gateway` resource previously shown in this procedure. The specification defines rules to route traffic with the `/reviews/` path prefix to the reviews service in the `bookinfo` namespace. The `VirtualService` explicitly references the gateway to ensure that the rules are only applied to the traffic that enters through the specified gateway.

Author: done

In the following example an {istio} `Gateway` resource configures a gateway proxy to act as an entry point for external traffic. This configuration exposes port 443 (HTTPS) for the host, `bookinfo.com`. The example configuration applies to pods with the `istio: ingressgateway` label. These pods typically correspond to ingress gateway pods. The `tls` mode is configured as `SIMPLE`, which terminates the incoming HTTPS traffic using the certificate and private key the example provides.

.Sample configuration
[source,yaml,subs="attributes,verbatim"]
----
apiVersion: networking.istio.io/v1
kind: Gateway
metadata:
  name: bookinfo-gateway
  namespace: bookinfo
spec:
  selector:
    istio: ingressgateway
  servers:
  - port:
      number: 443
      name: https-443
      protocol: HTTPS
    hosts:
    - bookinfo.com
    tls:
      mode: SIMPLE
      serverCertificate: /etc/certs/servercert.pem
      privateKey: /etc/certs/privatekey.pem
# ...
----

> The following `VirtualService` is bound to the {istio} `Gateway` resource previously shown in the previous example configuration. The specification defines rules to route traffic with the `/reviews/` path prefix to the reviews service in the `bookinfo` namespace. The `VirtualService` explicitly references the gateway to ensure that the rules are only applied to the traffic that enters through the specified gateway.

Reviewer 3: Not worth holding the merge, but "previously shown in the previous example" seems redundant.

```suggestion
The following `VirtualService` is bound to the {istio} `Gateway` resource shown in the previous example configuration. The specification defines rules to route traffic with the `/reviews/` path prefix to the reviews service in the `bookinfo` namespace. The `VirtualService` explicitly references the gateway to ensure that the rules are only applied to the traffic that enters through the specified gateway.
```
